comp/core/workloadmeta/collectors/internal/ecs/v4parser.go

```diff
@@ -81,18 +84,54 @@ func (c *collector) getTaskWithTagsFromV4Endpoint(ctx context.Context, task v1.T
 		return v1TaskToV4Task(task), errors.New(err)
 	}
 
-	taskWithTags, err := c.metaV3or4(metaURI, "v4").GetTaskWithTags(ctx)
-	if err != nil {
-		// If it's a timeout error, log it as debug to avoid spamming the logs as the data can be fetched in next run
-		if errors.Is(err, context.DeadlineExceeded) {
-			log.Debugf("timeout while getting task with tags from metadata v4 API: %s", err)
-		} else {
-			log.Warnf("failed to get task with tags from metadata v4 API: %s", err)
+	var returnTask v3or4.Task
+	// If we have tags, are collecting tags and have zero tags present in the cache
+	// Call getTaskWithTags
+	if c.hasResourceTags && c.collectResourceTags && len(rt.tags) == 0 && len(rt.containerInstanceTags) == 0 {
+		// No tags in the cache, fetch from metadata v4 API
+		taskWithTags, err := c.metaV3or4(metaURI, "v4").GetTaskWithTags(ctx)
+		if err != nil {
+			// If it's a timeout error, log it as debug to avoid spamming the logs as the data can be fetched in next run
+			if errors.Is(err, context.DeadlineExceeded) {
+				log.Debugf("timeout while getting task with tags from metadata v4 API: %s", err)
+			} else {
+				log.Warnf("failed to get task with tags from metadata v4 API: %s", err)
+			}
+			return v1TaskToV4Task(task), err
+		}
+
+		// Add tags to the cache
+		rt = resourceTags{
+			tags:                  taskWithTags.TaskTags,
+			containerInstanceTags: taskWithTags.ContainerInstanceTags,
 		}
-		return v1TaskToV4Task(task), err
+
+		c.resourceTags[task.Arn] = rt
+		returnTask = *taskWithTags
+	} else {
+		// Do not query for tags, but use them if they exist
+		taskWithTags, err := c.metaV3or4(metaURI, "v4").GetTask(ctx)
+		if err != nil {
+			// If it's a timeout error, log it as debug to avoid spamming the logs as the data can be fetched in next run
+			if errors.Is(err, context.DeadlineExceeded) {
+				log.Debugf("timeout while getting task from metadata v4 API: %s", err)
+			} else {
+				log.Warnf("failed to get task from metadata v4 API: %s", err)
+			}
+			return v1TaskToV4Task(task), err
+		}
+		// Add tags if they exist
+		if len(rt.tags) == 0 {
```

Comment:
should be `if len(rt.tags) != 0 {` ?

Reply:
Good catch on this

Reply:
Added missing test and corrected test data